Bolt address with same origin as http address

maxsumrall opened the pull request:
SharedServerTestBase should ultimately not have those changes.

chrisvest committed:
Make all test databases - impermanent and embedded alike - use 8 MiB page cache memory by default

Databases created with the TestGraphDatabaseFactory now all use 8 MiB of page cache memory by default, unless explicitly configured otherwise.
This setting previously only applied to impermanent databases.
Now it applies to embedded databases as well.

apcj committed:
Send only a batch of transactions in response to each pull request

apcj committed:
Cope with batched transactions in catchup.

apcj commented on community/server-api/src/main/java/org/neo4j/server/rest/repr/Serializer.java:
```diff
@@ -76,6 +76,11 @@ final void serialize( ListWriter list, ListRepresentation value )
     }
 
+    final String relativeBoltUri(String boltPort){
+        String bolt = baseUri.toString().replaceFirst( baseUri.getScheme(), "bolt" );
```
This seems a bit backward to me. Why throw away the parsed URI and use string manipulation (replaceFirst, length etc.) when when you can neatly extract the host part out of the URI with a method call?

maxsumrall replied:
Yes, I had also found that to be a better way to do this.

apcj commented on community/server/src/test/java/org/neo4j/server/rest/DiscoveryServiceIT.java:
```diff
@@ -83,14 +83,39 @@ public void shouldRedirectOnHtmlRequest() throws Exception
+        // then
+        Map<String,Object> map = JsonHelper.jsonToMap( response.getEntity() );
+        assertThat( String.valueOf( map.get( "bolt" ) ),
+                containsString( "bolt://" + request.getHeaders().get("host") ) );
```
This doesn't seem like a very strong assertion. For example, does the URI have the correct port?

maxsumrall replied:
Yes I'll amend it to also check for the port. The main concern was to test that the host part of the address was correct.

pe4cey committed:
Bumped version of Neo4j Browser to 2.0.0-M07

Commit:
Respond with bolt discovery address as the host of the incoming request, like http.

chrisvest commented:
@maxsumrall Looks like some extra commits snuck in by accident somehow.

jimwebber commented:
Following some conversation with @benbc we're going to affine this with the existing Bolt advertised address stuff such that an explicitly configured address has precedence over an inferred address. Until that's pushed, this remains not ready for merge.

jimwebber committed:
Making Bolt advertised address affined to HTTP host

This is a belt-n-braces style commit that tries to address some of the
inconveniences of Bolt server being a kernel extension.
If no sane advertised address is specified for Bolt, then the discovery
service will now try to figure one out using the HTTP Host header
as the basis for computing a Bolt connection URI.

This embeds some assumptions around advertised addresses that may
not be true in all possible cases but should be true for the vast majority.
In particular if the server is bound to a different NIC for HTTP and Bolt,
it's likely this won't work as expected.

jimwebber committed:
Making Bolt advertised address affined to HTTP host

This is a belt-n-braces style commit that tries to address some of the
inconveniences of Bolt server being a kernel extension.
If no sane advertised address is specified for Bolt, then the discovery
service will now try to figure one out using the HTTP Host header
as the basis for computing a Bolt connection URI.

This embeds some assumptions around advertised addresses that may
not be true in all possible cases but should be true for the vast majority.
In particular if the server is bound to a different NIC for HTTP and Bolt,
it's likely this won't work as expected.

benbc commented:
Fine by me.
